File: Euclid/Game/Permissions/UserGroup.cs
using System.Collections.Generic;

namespace Euclid.Game
{
    public class UserGroup
    {
        #region Properties

        public string Name { get; set; }
        public int Rank { get; set; }
        public List<string> DefaultPermissions { get; set; }
        public List<string> InheritsGroups { get; set; }
        public List<string> ExcludesPermissions { get; set; }
        public HashSet<string> Permissions { get; set; }

        #endregion

        #region Constructor

        public UserGroup(string name, int rank)
        {
            Name = name;
            Rank = rank;
            DefaultPermissions = new List<string>();
            InheritsGroups = new List<string>();
            ExcludesPermissions = new List<string>();
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Build permissions by taking into account the default perms, the inherits and excludes.
        /// </summary>
        public void BuildPermissions(UserGroup[] groups)
        {
            Permissions = new HashSet<string>();

            foreach (var permission in DefaultPermissions)
                Permissions.Add(permission);

            foreach (var userGroup in groups)
                foreach (var inheritGroup in InheritsGroups)
                {
                    if (inheritGroup == userGroup.Name)
                    {
                        foreach (var permission in userGroup.DefaultPermissions)
                            Permissions.Add(permission);
                    }
                }

            foreach (var excludePermission in ExcludesPermissions)
                Permissions.Remove(excludePermission);
        }

        /// <summary>
        /// Get if the group has permission
        /// </summary>
        public bool HasPermission(string permission)
        {
            if (Permissions.Contains("root"))
                return true;

            return Permissions.Contains(permission);
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/permissions/UserGroup.java
package euclid.game.permissions;

import java.util.HashSet;
import java.util.List;

public class UserGroup {

    private String name;
    private int rank;
    private List<String> defaultPermissions;
    private List<String> inheritsGroups;
    private List<String> excludesPermissions;
    private HashSet<String> permissions;

    public UserGroup(String name, int rank) {
        this.name = name;
        this.rank = rank;
        this.defaultPermissions = new java.util.ArrayList<>();
        this.inheritsGroups = new java.util.ArrayList<>();
        this.excludesPermissions = new java.util.ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getRank() {
        return rank;
    }

    public void setRank(int rank) {
        this.rank = rank;
    }

    public List<String> getDefaultPermissions() {
        return defaultPermissions;
    }

    public void setDefaultPermissions(List<String> defaultPermissions) {
        this.defaultPermissions = defaultPermissions;
    }

    public List<String> getInheritsGroups() {
        return inheritsGroups;
    }

    public void setInheritsGroups(List<String> inheritsGroups) {
        this.inheritsGroups = inheritsGroups;
    }

    public List<String> getExcludesPermissions() {
        return excludesPermissions;
    }

    public void setExcludesPermissions(List<String> excludesPermissions) {
        this.excludesPermissions = excludesPermissions;
    }

    public HashSet<String> getPermissions() {
        return permissions;
    }

    public void setPermissions(HashSet<String> permissions) {
        this.permissions = permissions;
    }

    public void buildPermissions(UserGroup[] groups) {
        permissions = new HashSet<>();

        for (String permission : defaultPermissions)
            permissions.add(permission);

        for (UserGroup group : groups) {
            for (String inheritGroup : inheritsGroups) {
                if (inheritGroup.equals(group.getName())) {
                    for (String permission : group.getDefaultPermissions())
                        permissions.add(permission);
                }
            }
        }

        for (String excludePermission : excludesPermissions)
            permissions.remove(excludePermission);
    }

    public boolean hasPermission(String permission) {
        if (permissions.contains("root"))
            return true;

        return permissions.contains(permission);
    }
}
